src/nools/lib.js

```diff
@@ -140,40 +140,32 @@ function emitTasksForSchedule(c, schedule, r) {
   }
 }
 
-function emitContactBasedTargetFor(c, targetConfig) {
-  if(targetConfig.appliesIf && !targetConfig.appliesIf(c)) return;
+function emitTargetFor(targetConfig, c, r) {
+  var isEmittingForReport = !!r;
+  if(targetConfig.appliesIf && !targetConfig.appliesIf(c, r)) return;
 
-  var pass = !targetConfig.passesIf || !!targetConfig.passesIf(c);
+  var pass = !targetConfig.passesIf || !!targetConfig.passesIf(c, r);
+  var instanceDoc = isEmittingForReport ? r : c.contact;
+  var instance = createTargetInstance(targetConfig.id, instanceDoc, pass);
+
+  instance._id = (targetConfig.idType === 'report' ? r && r._id : c.contact._id) + '~' + targetConfig.id;
 
-  var instance = createTargetInstance(targetConfig.id, c.contact, pass);
   if(typeof targetConfig.date === 'function') {
-    instance.date = targetConfig.date(c);
+    instance.date = targetConfig.date(c, r);
   } else if(targetConfig.date === undefined || targetConfig.date === 'now') {
     instance.date = now.getTime();
   } else if(targetConfig.date === 'reported') {
-    instance.date = c.reported_date;
+    instance.date = isEmittingForReport ? r.reported_date : c.contact.reported_date;
   } else {
     throw new Error('Unrecognised value for target.date: ' + targetConfig.date);
   }
-  emitTargetInstance(instance);
-}
 
-function emitReportBasedTargetFor(c, r, targetConf) {
-  var instance, pass;
-  if(targetConf.appliesIf && !targetConf.appliesIf(c, r)) return;
-
-  if(targetConf.emitCustom) {
-    targetConf.emitCustom(c, r);
+  if(targetConfig.emitCustom) {
+    targetConfig.emitCustom(instance, c, r);
```

Comment:
You're changing the parameter order for `emitCustom`, wouldn't that break existent configurations? 
It also differs from the the documented version: https://github.com/medic/medic-docs/blob/master/configuration/developing-community-health-applications.md#target-properties

Reply:
Looking at the issue, the param order you proposed was compatible with existent configurations.
```
emitCustom: (contact, report, instance) => {
```

Reply:
Ya. I am making breaking interface changes here. 

The reason I'm doing it like this is because `report` isn't always defined. It's awkward to document that the signature of the function is `emitCustom(c, instance)` for scenario one and `emitCustom(c, r, instance)` for scenario two. So I'd like to move all optional parameters to the end.